File: pipelines/processors/enhancement_processor.py
import cv2
import logging
import numpy as np
from domain.exceptions import ProcessorError
from pipelines.processors.base_processor import BaseProcessor

logger = logging.getLogger(__name__)


class EnhancementProcessor(BaseProcessor):
    name = "enhancement"

    def __init__(self):
        super().__init__(
            transform_steps = [
                self._denoise,
                self._apply_clahe,
                self._sharpen,
            ]
        )


    def _denoise(self, image, context: dict) -> np.ndarray:
        denoise_h = context.get("denoise_h", 10)
        template_window_size = context.get("denoise_template_window_size", 7)
        search_window_size = context.get("denoise_search_window_size", 21)

        try:
            result = cv2.fastNlMeansDenoising(
                image,
                None,
                denoise_h,
                template_window_size,
                search_window_size
            )
        except Exception as e:
            logger.error(f"Error when applying denoise: {e}")
            raise ProcessorError(f"Error when applying denoise: {e}") from e

        return result


    def _apply_clahe(self, image, context: dict) -> np.ndarray:
        clip_limit = context.get("clahe_clip_limit", 2.0)
        tile_grid_size = context.get("clahe_tile_grid_size", (4, 4))

        clahe = cv2.createCLAHE(clipLimit=clip_limit, tileGridSize=tile_grid_size)

        try:
            result = clahe.apply(image)
        except Exception as e:
            logger.error(f"Error when applying CLAHE: {e}")
            raise ProcessorError(f"Error when applying CLAHE: {e}") from e

        return result


    def _sharpen(self, image, context: dict) -> np.ndarray:
        strength = context.get("sharpen_strength", 1.2)
        blur_ksize = context.get("sharpen_blur_kernel_size", (0, 0))
        sigma = context.get("sharpen_sigma", 1.0)

        try:
            blurred = cv2.GaussianBlur(image, blur_ksize, sigma)
        except Exception as e:
            logger.error(f"Error when blurring during sharpen: {e}")
            raise ProcessorError(f"Error when blurring during sharpen: {e}") from e

        try:
            result = cv2.addWeighted(image, 1.0 + strength, blurred, -strength, 0)
        except Exception as e:
            logger.error(f"Error when applying sharpen: {e}")
            raise ProcessorError(f"Error when applying sharpen: {e}") from e

        return result

File: cv/processors/enhancement_processor.go
package processors

import (
	"fmt"
	cv "gocv.io/x/gocv"
	"image"
	m "tyre-match-backend/db/models"
)

type EnhancementProcessor struct {
	BaseProcessor

	DenoiseH                  float32
	DenoiseTemplateWindowSize int
	DenoiseSearchWindowSize   int

	BlurSigma      float64
	BlurKSize      image.Point
	BlurBorderType cv.BorderType

	SharpenStrength float64

	CLAHEClipLimit    float64
	CLAHETileGridSize image.Point
}

func NewEnhancementProcessor() *EnhancementProcessor {
	enhancementProcessor := &EnhancementProcessor{
		BaseProcessor: BaseProcessor{
			Name:     "enhancement",
			FileType: m.FileTypeEnhanced,
		},

		DenoiseH:                  10,
		DenoiseTemplateWindowSize: 7,
		DenoiseSearchWindowSize:   21,

		BlurSigma:      1.0,
		BlurKSize:      image.Pt(0, 0),
		BlurBorderType: cv.BorderDefault,

		SharpenStrength: 1.2,

		CLAHEClipLimit:    3.0,
		CLAHETileGridSize: image.Pt(4, 4),
	}

	enhancementProcessor.ProcessingSteps = []ProcessingStep{
		enhancementProcessor.Denoise,
		enhancementProcessor.ApplyCLAHE,
		//enhancementProcessor.Sharpen,
	}

	return enhancementProcessor
}

func (p *EnhancementProcessor) ApplyCLAHE(source, destination *cv.Mat) error {
	err := p.ValidateSourceImage(source)
	if err != nil {
		return fmt.Errorf("apply clahe %v", err)
	}

	clahe := cv.NewCLAHEWithParams(p.CLAHEClipLimit, p.CLAHETileGridSize)
	err = clahe.Apply(*source, destination)
	if err != nil {
		return fmt.Errorf("apply clahe - apply failed: %v", err)
	}

	return nil
}

func (p *EnhancementProcessor) Denoise(source, destination *cv.Mat) error {
	err := p.ValidateSourceImage(source)
	if err != nil {
		return fmt.Errorf("denoise %v", err)
	}

	err = cv.FastNlMeansDenoisingWithParams(*source, destination, p.DenoiseH, p.DenoiseTemplateWindowSize, p.DenoiseSearchWindowSize)
	if err != nil {
		return fmt.Errorf("denoise - fast nl means denoising failed: %v", err)
	}

	return nil
}

func (p *EnhancementProcessor) Sharpen(source, destination *cv.Mat) error {
	err := p.ValidateSourceImage(source)
	if err != nil {
		return fmt.Errorf("sharpen %v", err)
	}

	blurred := cv.NewMat()
	defer blurred.Close()

	err = cv.GaussianBlur(*source, &blurred, p.BlurKSize, p.BlurSigma, p.BlurSigma, p.BlurBorderType)
	if err != nil {
		return fmt.Errorf("sharpen - gaussian blur failed: %v", err)
	}
	if blurred.Empty() {
		return fmt.Errorf("sharpen - gaussian blur produced an empty image")
	}

	err = cv.AddWeighted(*source, 1.0+p.SharpenStrength, blurred, -p.SharpenStrength, 0, destination)
	if err != nil {
		return fmt.Errorf("sharpen - add weighted failed: %v", err)
	}
	if destination.Empty() {
		return fmt.Errorf("sharpen - add weighted produced an empty image")
	}

	return nil
}
